src/theory/theory_inference.h

```diff
@@ -0,0 +1,102 @@
+/*********************                                                        */
+/*! \file theory_inference.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The theory inference utility
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__THEORY_INFERENCE_H
+#define CVC4__THEORY__THEORY_INFERENCE_H
+
+#include "context/cdhashmap.h"
+#include "expr/node.h"
+#include "theory/output_channel.h"
+
+namespace CVC4 {
+namespace theory {
+
+class TheoryInferenceManager;
+
+/**
+ * A theory inference base class. This class is an abstract data structure for
+ * storing pending lemmas or facts in the buffered inference manager. It can
+ * be seen a single use object capturing instructions for making a single
+ * call to TheoryInferenceManager for lemmas or facts.
+ */
+class TheoryInference
+{
+ public:
+  virtual ~TheoryInference() {}
+  /**
+   * Called by the provided inference manager to process this inference. This
+   * method should make a single call to inference manager to process this
+   * inference (e.g. a call to trustedLemma, assertFactInternal etc.).
```

Comment:
I have two issues with this comment:
1. Why only a single call? We explicitly want to do multiple calls here: one to push this lemma to the output channel and one to do side effects, for example.
2. While I agree with the overall decision, I expect the fact that the lemma needs to push itself to the output channel in a callback called by the inference manager to be somewhat counter intuitive to the average user. I'd thus try to stress that this callback needs to do this.

Reply:
For 1, I'll say this is the typical (but not required) use case.